Close a WebSocketTask if it completes with an error without closing properly

achristensen07 opened the pull request:
#### ebe1ce92f75fb6c7eac7fc2002af43df30723be7
<pre>
Close a WebSocketTask if it completes with an error without closing properly
<a href="https://bugs.webkit.org/show_bug.cgi?id=274003">https://bugs.webkit.org/show_bug.cgi?id=274003</a>
<a href="https://rdar.apple.com/110487778">rdar://110487778</a>

Reviewed by Brady Eidson.

If the TCP connection has an error but we have not received a close frame, then
CFNetwork will call didCompleteWithError but won&apos;t call the completion block of
receiveMessageWithCompletionHandler.  This is strange, but to make our interface
with CFNetwork more robust, close the WebSocket in this case if it hasn&apos;t already
been closed.  WebSocketTask::didClose will ignore this call if it has already
been closed.

Unfortunately, the API doesn&apos;t exist to make a unit test for this.  In attempting
to make a unit test, I filed <a href="https://rdar.apple.com/127883777">rdar://127883777</a> which, when resolved will allow us
to test cases similar to this.  That radar shows that future work is still to be
done in this area, but manual testing verifies this fixes the issue in the
original radar of <a href="https://rdar.apple.com/110487778">rdar://110487778</a>

There are 3 ways a WebSocketTask can close:
1. receiveMessageWithCompletionHandler completion block can be called with an error
2. URLSession:webSocketTask:didCloseWithCode:reason: can be called
3. URLSession:task:didCompleteWithError: can be called
It is an error for didCompleteWithError to be called before another call has closed
the WebSocket, so this change calling didClose whether error is null or not is correct.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):

Canonical link: <a href="https://commits.webkit.org/278623@main">https://commits.webkit.org/278623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03656a502f2d6b2c955d76302da27f650860fe1d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30437 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3468 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54392 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1825 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36731 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1499 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41643 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53234 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44073 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22759 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25396 "Passed tests") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47376 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55988 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26245 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1294 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49041 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27493 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44140 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48175 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28374 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7429 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -899,6 +900,8 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
     if (auto networkDataTask = [self existingTask:task])
         networkDataTask->didCompleteWithError(error, networkDataTask->networkLoadMetrics());
+    else if (auto* webSocketTask = [self existingWebSocketTask:task])
+        webSocketTask->didClose(WebCore::ThreadableWebSocketChannel::CloseEventCodeAbnormalClosure, emptyString());
```
On the surface, it seemed weird to me that we might tell the existing web socket task there's an error (`CloseEventCodeAbnormalClosure`) even if there's no `NSError` passed in here. (e.g. it's a "normal" completion from the NSURLSession standpoint)

In discussion with Alex, he mentioned it's an error for a web socket to end without having sent a close frame, and that's what's captured here.

I'd like a comment here to that effect to make that crystal clear to future code spelunkers.

achristensen07 committed:
Close a WebSocketTask if it completes with an error without closing properly
https://bugs.webkit.org/show_bug.cgi?id=274003
rdar://110487778

Reviewed by Brady Eidson.

If the TCP connection has an error but we have not received a close frame, then
CFNetwork will call didCompleteWithError but won't call the completion block of
receiveMessageWithCompletionHandler.  This is strange, but to make our interface
with CFNetwork more robust, close the WebSocket in this case if it hasn't already
been closed.  WebSocketTask::didClose will ignore this call if it has already
been closed.

Unfortunately, the API doesn't exist to make a unit test for this.  In attempting
to make a unit test, I filed rdar://127883777 which, when resolved will allow us
to test cases similar to this.  That radar shows that future work is still to be
done in this area, but manual testing verifies this fixes the issue in the
original radar of rdar://110487778

There are 3 ways a WebSocketTask can close:
1. receiveMessageWithCompletionHandler completion block can be called with an error
2. URLSession:webSocketTask:didCloseWithCode:reason: can be called
3. URLSession:task:didCompleteWithError: can be called
It is an error for didCompleteWithError to be called before another call has closed
the WebSocket, so this change calling didClose whether error is null or not is correct.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):

Canonical link: https://commits.webkit.org/278623@main
